### ISSUE TITLE
Line number (position information) not available in `lark.Tree`

### ISSUE THREAD

**Issue by clouds56** [lark-parser/lark#82]
After performing `Transformer`, the position information in `Token` is lost.
It would be convenient to have the information also in `Tree`.

BTW, adding an API that return the original text of a `Tree` is also useful.

**Comment by erezsh** [comment]
Hi, Lark has the option to `propagate_positions`:

   parser = Lark(..., propagate_positions=True)

It propagate the line/column information into the tree nodes, all the way to the root. Hopefully that solves your issue?

You could use the information in conjunction with the original text to retrieve it per tree. I'm reluctant to add it as an official API because I don't think it's a common use-case, and also because it wouldn't work for text streams, which Lark fully supports.

**Comment by clouds56** [comment]
Yes, `propagate_positions` is exactly what I need, haven't noticed this option before.
For the original text API that's fine, if I have the position information, I could easily get the original text on my own.

However the `tree.line` and `tree.column` doesn't work well with `Transformer`
The position information is not available when children are all transformed, while the head node of the subtree is not passed into the transform function.

I made an workaround by modify https://github.com/erezsh/lark/blob/master/lark/tree.py#L121 to `return f(tree, *items)`.

Is it possible to accept some patch like this workaround?
If considering backward compatibility, we could have an option to switch between `return f(tree, *items)` and `return(items)`.
I prefer expand `items` to `*items` since I have things like `[func_name, *params] = items` in almost all method in Transformer. And if someone like the old way, it could easily use
```
def subtree(tree, *items):
    pass
```
in the transformer.

**Comment by erezsh** [comment]
I've wanted to reform the transformers for quite some time, so I will welcome a solution that addresses every need. But, it's not very simple.

The reason for the original interface (of `rulename(self, items)`) is that when providing the LALR parser with a transformer, it doesn't have to construct a tree node, which improves performance.

And of course, it's also more convenient in the general case. But sometimes passing `*items` is more convenient, and sometimes passing `tree` is necessary for line information.

I still need to figure out how can these 3 interfaces live together in a way that makes sense.

**Comment by night199uk** [comment]
@clouds56 I hit the same issue a few weeks ago. I raised #103 as a temporary stopgap back then, but I think @erezsh has something bigger in mind. You could possibly pull that PR as an interim stopgap - but it sounds like you have something already?

**Comment by erezsh** [comment]
I currently have two ideas to address this:

1) The default transformer will accepts a `tree` argument, instead of items. So line information (and other attributes) will be accessible. You can then use the `items` decorator (or also `inline_items`) to modify a specific method, or the entire transformer. Then, the LALR parser can require a decorated transformer.

2) The default transformer will change the trees in-place, which means that attributes (like line info) will be preserved. There will also be a CopyTransformer to provide the old functionality.

Possibly, I'll try to combine both somehow. I'm still not sure how to do it in a way that's simple to use.

**Comment by night199uk** [comment]
I think I prefer 1). This makes Transformers consistent with Visitors, which already seem to take a tree as an argument instead of a list of children. Second, this makes it easier to return identity (return tree) in the Transformers if we decide after looking at it that we don't want to transform a particular item.

Since you're passing in a reference to the Tree, I think this would implicitly allows 2) anyway, right? If desired, the tree can be modified in-place since it's a reference to the original item?

**Comment by night199uk** [comment]
Re: InlineTransformers though, one other thought. How about moving the tree metadata (column, line, etc) into a dict 'metadata' property or so on the tree? Then, when inlining args, the metadata dict can be passed as **kwargs. So you end up with (pseudocode):
f(*tree.children, **tree.metadata).

Not sure if there is a way to do this without requiring all InlineTransformer functions add the **kwargs argument to absorb unused kwargs. that might require some more thought too.

**Comment by clouds56** [comment]
@night199uk I like the idea that split metadata to another dict.
While I prefer pass the metadata via `f(*tree.children, meta=tree.metadata)`, and thus metadata could be a class other than dict.
As I said before, have an option on this, user could easily choose between `f(*tree.children)` and the metadata version.
Alternatively we could use [inspect.signature](https://docs.python.org/3/library/inspect.html#inspect.signature) to detect if `f` have parameter `meta` and decide if we pass the metadata

```python
import inspect
def _get_func(f):
    f = getattr(self, name)
    has_meta = 'meta' in inspect.signature(f).parameters
    return has_meta, f
```

**Comment by erezsh** [comment]
@night199uk  I prefer (1) as well.  And I like the idea of moving line info to a `meta` attribute. It's backwards incompatible, but that's fine, and I can change the transformers at the same time.

The thing about (2), is that after you recurse into the children, you need to either provide the user with a *new* tree instance, or change the existing one. But that's not a big deal anymore if we use the `meta` attribute. So I can keep the immutable transformer as default (which is more intuitive), and also provide an "InplaceTransformer" (or a better name?) for convenience and a bit of speed, if anyone needs it.

@clouds56 @night199uk 

That can be decided in the decorator. I.e. `@inline_items(meta=True)` or `@items(meta='auto')`, and the latter can use the inspect trick (assuming it plays nice with the rest).

So what do you guys think, do we have a plan?

**Comment by erezsh** [comment]
@night199uk @clouds56 

I spent some time thinking on how to best design the new transformers, trying to strike the right balance between simplicity, functionality and verbosity.

After moving all the line information into `meta`, I realized that there's no scenario in which I need to pass the tree. Everything I need is in `children` and `meta`. So the existing Transformer interface can stay relatively the same as it is now, but with an option to add `meta`. Do you agree with my conclusion?

Or do you think it's better to pass the tree anyway for some reason?

**Comment by night199uk** [comment]
Hey, sorry - haven’t had a chance to look at this in a bit. Will be a week or two but will cycle back through and test out he new implementation then. Thanks :-)

**Comment by erezsh** [comment]
Fixed in 0.6.0
